### ISSUE TITLE
Move org.eclipse.equinox.internal.p2.repository.Credentials to org.eclipse.equinox.p2.transport.ecf

### ISSUE THREAD

**Issue by laeubi** [eclipse-equinox/p2#200]
The `Credentials` class is currently only used in the `RepositoryTransport` and imposes a requirement to equinox.security in p2.repository. If we can move this, the dependency footprint of p2.repository. can be reduced.

**Comment by akurtakov** [comment]
Is Credentials API or not?

**Comment by laeubi** [comment]
> Is Credentials API or not?

From the package name it is internal:

org.eclipse.equinox.**internal**.p2.repository.Credentials

**Comment by akurtakov** [comment]
I got it now so the requirement on equinox.security is transitive one. As this is internal class with single usage it sounds good to move it where it's used.

**Comment by merks** [comment]
For doing useful things with p2, e.g., accessing repositories that require credentials, calling the following method is useful, which we do in Oomph for the targlet implementation:

org.eclipse.equinox.internal.p2.repository.Credentials.forLocation(URI, boolean)

Keep in mind that this is "is it API question" is double-edged sword that cuts two ways, especially for p2, where the API is so poor you can't do much of anything without dropping down.  After all, I too can change anything I like simply to suit Oomph or to suit any other goal that I **personally** have as a priority and I can do so with no regard to the impact on Tycho or anyone thing because "is it API" is typically answered with no.  In the past, I've simply tried to leave p2 alone and get by with what's there to be minimally disruptive, but that not actually necessary.  I expect there isn't a double standard in this regard.

It's mostly a rhetorical question in this specific case because in this case I can no doubt use reflection to call that method, regardless of where the class is located and if this helps Tycho then I can compromise for the good of the many.  But I do ask, in general, for a little more regard than "is it API" when it comes to p2 which is heavily used by Oomph.

**Comment by laeubi** [comment]
> calling the following method is useful, which we do in Oomph for the targlet implementation:
> 
> org.eclipse.equinox.internal.p2.repository.Credentials.forLocation(URI, boolean)

Just to be clear, the class won't vanish or change, it will just move to another p2-bundle... if it helps we can even choose the same package name, so if both bundles are imported one don't even need to change the code.

I think the biggest problem of P2 is that no one dare to change anything and instead we have numerous workaround, re implementations and possibly even duplicate efforts. So lots of cool stuff in Oompf can only be there any neither Tycho nor platfrom can befit from it, maybe there is also useful stuff in Tycho (e.g. `PGPAdvice`) ... so my goal is to possibly migrate as much as possible from Tycho > P2 over time.

And i can say from the Tycho POV if P2 has breaking changes for good reasons Tycho simply will adapt. But adding new API sadly always seems not valuable because this puts a real high burden on evolution and innovation of such code even if it is claimed to be a keypoint of "eclipse-platform" I today try to do whatever is possible on internal stuff because all "public-API" is simply almost always not worth the effort to put into as it never pay-back anything for my use-cases where I try to adapt to changed/new code instead of trying to stay back as much as possible.

I think I mentioned it already elsewhere, for me a release is wasted effort if "nothing changes"!

**Comment by merks** [comment]
Yes, I know the class will just move.  Maybe it has the same package name or maybe not.  Unfortunately split packages also have a cost/burden.   If you import a package, you probably end up depending on all the bundles that export the package (depending on uses clauses or stuff like I suppose).  Also, when a new certificate is issued, you have to be sure that both bundles use the same certificate.   Whichever way it goes, I'll try to accommodate.

When p2 breaks API, or provides new API I must use, of course Oomph can accommodate that.  But then I have to basically change all of Oomph to be installable only in the latest platform version or end up with very nasty problems like this one:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=579338

That's a major pain in the butt...

So yes, while we should define API, we're not made of endless time **and** that's a huge investment that in itself becomes a bigger burden over time.

**Comment by laeubi** [comment]
Yes right split packages are really bad so we should avoid this, this was just if there is a demand to have the package stay the same.

It is just not clear to me if you think that if we move the class if it will create any problem for Oomph? I assume Oompf uses ECF-Transport so it would just be a matter of adjust the import? Maybe it would also be good to know how Ooompf is using that method, I would assume that this class is only used internally in P2 an never needs to be used by client code.

For example if you want to inject credentials (fetching them would work the same) Tycho has had the following Util method:
https://github.com/eclipse-tycho/tycho/blob/tycho-3.0.x/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/helper/P2PasswordUtil.java

Or is Oompf doing the authetication differently? Maybe then we can better "open" some code in the Traansportmsoo Oompf is able to reuse what ECF does...

**Comment by merks** [comment]
Oomph (like the Platform, Tycho, m2e, and so many other projects) has a setup that you can use to easily see what it uses.

https://www.eclipse.org/setups/installer/?url=http://git.eclipse.org/c/oomph/org.eclipse.oomph.git/plain/setups/configurations/OomphConfiguration.setup&show=true

Yes of course I can use the new package name, but then I cannot install Oomph anywhere but in 4.27 and I would prefer not to do that, and certainly not for something so far off the beaten path, i.e., handling authenticated repositories while generating a settings.xml file.

**Comment by merks** [comment]
So as long as the class is somewhere and I can call that method on it, I'll just modify the code to invoke things reflectively so that it works in  < 4.27 and >= 4.27...
